README.md
# heatshrink

A data compression/decompression library for embedded/real-time systems.


## Key Features:

- **Low memory usage (as low as 50 bytes)**
    It is useful for some cases with less than 50 bytes, and useful
    for many general cases with < 300 bytes.
- **Incremental, bounded CPU use**
    You can chew on input data in arbitrarily tiny bites.
    This is a useful property in hard real-time environments.
- **Can use either static or dynamic memory allocation**
    The library doesn't impose any constraints on memory management.
- **ISC license**
    You can use it freely, even for commercial purposes.


## Getting Started:

There is a standalone command-line program, `heatshrink`, but the
encoder and decoder can also be used as libraries, independent of each
other. To do so, copy `heatshrink_common.h`, `heatshrink_config.h`, and
either `heatshrink_encoder.c` or `heatshrink_decoder.c` (and their
respective header) into your project. For projects that use both,
static libraries are built that use static and dynamic allocation.

Dynamic allocation is used by default, but in an embedded context, you
probably want to statically allocate the encoder/decoder. Set
`HEATSHRINK_DYNAMIC_ALLOC` to 0 in `heatshrink_config.h`.

Note that the command line program defaults to compressing with default
parameters when run without any arguments -- to see usage info, use:

    $ heatshrink -h


### Basic Usage

1. Allocate a `heatshrink_encoder` or `heatshrink_decoder` state machine
using their `alloc` function, or statically allocate one and call their
`reset` function to initialize them. (See below for configuration
options.)

2. Use `sink` to sink an input buffer into the state machine. The
`input_size` pointer argument will be set to indicate how many bytes of
the input buffer were actually consumed. (If 0 bytes were consumed, the
buffer is full.)

3. Use `poll` to move output from the state machine into an output
buffer. The `output_size` pointer argument will be set to indicate how
many bytes were output, and the function return value will indicate
whether further output is available. (The state machine may not output
any data until it has received enough input.)

Repeat steps 2 and 3 to stream data through the state machine. Since
it's doing data compression, the input and output sizes can vary
significantly. Looping will be necessary to buffer the input and output
as the data is processed.

4. When the end of the input stream is reached, call `finish` to notify
the state machine that no more input is available. The return value from
`finish` will indicate whether any output remains. if so, call `poll` to
get more.

Continue calling `finish` and `poll`ing to flush remaining output until
`finish` indicates that the output has been exhausted.

Sinking more data after `finish` has been called will not work without
calling `reset` on the state machine.


## Configuration

heatshrink has a couple configuration options, which impact its resource
usage and how effectively it can compress data. These are set when
dynamically allocating an encoder or decoder, or in `heatshrink_config.h`
if they are statically allocated.

Note that the configuration parameters are not explicitly represented in
the compressed output -- in most use cases, either the configuration
settings will remain constant and can be implicit, or there will be some
kind of project-specific protocol envelope / header to contain that
information.


- `window_sz2`, `-w` in the CLI: Set the window size to 2^W bytes.

The window size determines how far back in the input can be searched for
repeated patterns. A `window_sz2` of 8 will only use 256 bytes (2^8),
while a `window_sz2` of 10 will use 1024 bytes (2^10). The latter uses
more memory, but may also compress more effectively by detecting more
repetition.

The `window_sz2` setting currently must be between 4 and 15.

- `lookahead_sz2`, `-l` in the CLI: Set the lookahead size to 2^L bytes.

The lookahead size determines the max length for repeated patterns that
are found. If the `lookahead_sz2` is 4, a 50-byte run of 'a' characters
will be represented as several repeated 16-byte patterns (2^4 is 16),
whereas a larger `lookahead_sz2` may be able to represent it all at
once. The number of bits used for the lookahead size is fixed, so an
overly large lookahead size can reduce compression by adding unused
size bits to small patterns.

The `lookahead_sz2` setting currently must be between 3 and the
`window_sz2` - 1.

- `input_buffer_size` - How large an input buffer to use for the
decoder. This impacts how much work the decoder can do in a single
step, and a larger buffer will use more memory. An extremely small
buffer (say, 1 byte) will add overhead due to lots of suspend/resume
function calls, but should not change how well data compresses.


### Recommended Defaults

For embedded/low memory contexts, a `window_sz2` in the 8 to 10 range is
probably a good default, depending on how tight memory is. Smaller or
larger window sizes may make better trade-offs in specific
circumstances, but should be checked with representative data.

The `lookahead_sz2` should probably start near the `window_sz2`/2, e.g.
-w 8 -l 4 or -w 10 -l 5. The command-line program can be used to measure
how well test data works with different settings.


## More Information and Benchmarks:

heatshrink is based on [LZSS], since it's particularly suitable for
compression in small amounts of memory. It can use an optional, small
[index] to make compression significantly faster, but otherwise can run
in under 100 bytes of memory. The index currently adds 2^(window size+1)
bytes to memory usage for compression, and temporarily allocates 512
bytes on the stack during index construction (if the index is enabled).

For more information, see the [blog post] for an overview, and the
`heatshrink_encoder.h` / `heatshrink_decoder.h` header files for API
documentation.

[blog post]: http://spin.atomicobject.com/2013/03/14/heatshrink-embedded-data-compression/
[index]: http://spin.atomicobject.com/2014/01/13/lightweight-indexing-for-embedded-systems/
[LZSS]: http://en.wikipedia.org/wiki/Lempel-Ziv-Storer-Szymanski


## Build Status

  [![Build Status](https://travis-ci.org/atomicobject/heatshrink.png)](http://travis-ci.org/atomicobject/heatshrink)


CHANGELOG.md
# heatshrink Changes By Release

## v0.4.2 - 2021-xx-yy

### API Changes

None.


### Bug Fixes

Fixed fallthrough in getopt 'h'.


### Other Improvements

Added `CHANGELOG.md`.

Added comment to help explain build error when attempting to build the
command line interface for embedded systems without getopt.h/err.h.

Added `#ifndef`s around defaults for static configuration variables
(`HEATSHRINK_STATIC_INPUT_BUFFER_SIZE`, `HEATSHRINK_STATIC_WINDOW_BITS`,
`HEATSHRINK_STATIC_LOOKAHEAD_BITS`, `HEATSHRINK_DEBUGGING_LOGS`,
`HEATSHRINK_USE_INDEX`). Thanks @ajaybhargav.

Use constant rather than bare int 4096 for buffer size.

Fixed fallthrough compiler warning.

Move repo into `include/`, `src/`, `test/`, and build in `build/`.

Update theft tests to match interface changes in theft.

greatest.h: Update vendored copy to 1.5.0.

Fix warnings with new version of greatest.

Fixed warnings for unreachable code on ARMCC. (Thanks @sport-climber.)

Fixed various typos, reported by @UnePierre.

Added checks and error messages when the CLI is given invalid
parameters for the window or lookahead bits.

Addressed a bunch of warnings from clang's `-Weverything`, which
mostly had to do with mixing up numeric types.


## v0.4.1 - 2015-11-01

### API Changes

None.


### Bug Fixes

Update inaccurate assertion. (Only active when logging.)


### Other Improvements

Fix minor typo.

benchmark: Scrape file size from ls | awk -- stat & du args are less portable.

Update contact info.


## v0.4.0 - 2015-06-07

### API Changes

The limits for the window and lookahead size options have changed. The
minimum lookahead bits option (-l) is now 3 (previously, 2), and the
maximum is the window bits - 1 (previously, equal to the window bits).

Change `get_bits()` to `uint16_t`.


### Bug Fixes

Fix search when window size equals lookahead.

Correct encoder state machine graph.


### Other Improvements

Fix signed/unsigned comparisons.

Use consistently signed vars during index construction.

Fix debug output and asserts.

Vary window and lookahead sizes in theft tests, add a regression test.
Thanks to @unixdj for finding this bug and providing very detailed
info about how to reproduce it!

Simplify break even point calculation a bit further.

Clean up / dead code elimination.

Misc test cleanup.

Add compression benchmarking script.

Initial work on a CONTRIBUTING.md.

README: Add details on example usage and recommended configuration defaults.

Consistently use "decompress", not "uncompress".


## v0.3.2 - 2015-01-03

### API Changes

None.


### Bug Fixes

Fix handling of `O_BINARY` on Windows.


### Other Improvements

Add optional property-based tests, using theft (if available).

Fix compilation on MinGW and other platforms without err.h

Print all status messages on stderr.

Fix missing #include for getopt.h.

Complete overhaul of Makefile: both static & dynamic variant static libs.



## v0.3.1 - 2014-06-25

### API Changes

None.


### Bug Fixes

Fix possible state yielding MORE from finish(), but 0 bytes from poll().


### Other Improvements

Improve non-indexed compression performance.

When indexed, only check matches that can beat the current best match.

Only check for matches that can be improvements.

Update heatshrink_common.h filename in README.md to reflect rename.

Add usage details, homepage to `heatshrink -h` message.

CLI: note that compression is the default.x


## v0.3.0 - 2013-12-28

### API Changes

Update non-indexed matching to use an `int16_t` pos.

Update index type to `int16_t`.

Decrease `HEATSHRINK_MAX_WINDOW_BITS` to 15.


### Bug Fixes

None.


### Other Improvements

Remove unused static library target from Makefile.

Cast `write(2)`'s -1 to (size_t) to avoid sign mismatch warning.

Code style cleanup: _always_ brace `if`s, extract NO_BITS constant.

Eliminate redundant comparison, index ensures first byte matches.

Improve throughput of `find_longest_match` via several improvements
to the indexing strategy.

If calling push_bits w/ 8 bits at beginning of new byte, just copy byte over whole.


## v0.2.1 - 2013-12-18

### API Changes

None.


### Bug Fixes

None.


### Other Improvements

Renamed `heatshrink.h` to `heatshrink_common.h`, to make it more
clearly separate from the `heatshrink.c` CLI interface.

Don't use assert unless `HEATSHRINK_DEBUGGING_LOGS` is set.

Fix some (potentially) unused variable and numeric type warnings.

Add warnings from `-Wextra`.

Add braces around statements hanging off `if`s that may compile out to no-ops.

Fix usage string ("[-v]" not "[v]").

Change from BSD-3 License to ISC License.

Reduce default `static_input_buffer_size`.

Update decoder state machine diagram to reflect MSB/LSB split.



## v0.2.0 - 2013-05-12

### API Changes

Change `uint16_t` to `size_t` for user-controlled buffer sizes.


### Bug Fixes

When index or count are > 8 bits long, ensure suspend/resume is correct.


### Other Improvements

Consistentcy: log to stderr for both enc and dec.

Tighten bounds for lookahead and window size.
Since they're only stored as `uint16_t`s, disallow overly large sizes.


## v0.1.0 - 2013-03-13

Initial public release.


src/heatshrink.c
#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>
#include <stdint.h>
#include <stdbool.h>
#include <assert.h>
#include <string.h>
#include <fcntl.h>

#include "heatshrink_encoder.h"
#include "heatshrink_decoder.h"

#define DEF_WINDOW_SZ2 11
#define DEF_LOOKAHEAD_SZ2 4
#define DEF_DECODER_INPUT_BUFFER_SIZE 256
#define DEF_BUFFER_SIZE (64 * 1024)

#if 0
#define LOG(...) fprintf(stderr, __VA_ARGS__)
#else
#define LOG(...) /* NO-OP */
#endif

/* If you're getting a build error because getopt.h or err.h are not
 * available when compiling for an embedded platform, you're probably
 * misundersting how this is supposed to be used.
 *
 * The encoder/decoder source files are for a library for embedded
 * devices, but this file is for a command line program for a Unix-like
 * environment, to work with data from embedded systems using the
 * heatshrink library, experiment with how effectively different
 * settings compress test data, etc. It is not supposed to be built for
 * embedded devices, which are unlikely to have those headers. */
#include <getopt.h>

#ifdef _WIN32
#include <errno.h>
#define HEATSHRINK_ERR(retval, ...) do { \
fprintf(stderr, __VA_ARGS__); \
fprintf(stderr, "Undefined error: %d\n", errno); \
exit(retval); \
} while(0)
#else
#include <err.h>
#define HEATSHRINK_ERR(...) err(__VA_ARGS__)
#endif

/*
 * We have to open binary files with the O_BINARY flag on Windows. Most other
 * platforms don't differentiate between binary and non-binary files.
 */
#ifndef O_BINARY
#define O_BINARY 0
#endif

static const int version_major = HEATSHRINK_VERSION_MAJOR;
static const int version_minor = HEATSHRINK_VERSION_MINOR;
static const int version_patch = HEATSHRINK_VERSION_PATCH;
static const char author[] = HEATSHRINK_AUTHOR;
static const char url[] = HEATSHRINK_URL;

static void usage(void) {
    fprintf(stderr, "heatshrink version %u.%u.%u by %s\n",
        version_major, version_minor, version_patch, author);
    fprintf(stderr, "Home page: %s\n\n", url);
    fprintf(stderr,
        "Usage:\n"
        "  heatshrink [-h] [-e|-d] [-v] [-w SIZE] [-l BITS] [IN_FILE] [OUT_FILE]\n"
        "\n"
        "heatshrink compresses or decompresses byte streams using LZSS, and is\n"
        "designed especially for embedded, low-memory, and/or hard real-time\n"
        "systems.\n"
        "\n"
        " -h        print help\n"
        " -e        encode (compress, default)\n"
        " -d        decode (decompress)\n"
        " -v        verbose (print input & output sizes, compression ratio, etc.)\n"
        "\n"
        " -w SIZE   Base-2 log of LZSS sliding window size\n"
        "\n"
        "    A larger value allows searches a larger history of the data for repeated\n"
        "    patterns, potentially compressing more effectively, but will use\n"
        "    more memory and processing time.\n"
        "    Recommended default: -w 8 (embedded systems), -w 10 (elsewhere)\n"
        "  \n"
        " -l BITS   Number of bits used for back-reference lengths\n"
        "\n"
        "    A larger value allows longer substitutions, but since all\n"
        "    back-references must use -w + -l bits, larger -w or -l can be\n"
        "    counterproductive if most patterns are small and/or local.\n"
        "    Recommended default: -l 4\n"
        "\n"
        " If IN_FILE or OUT_FILE are unspecified, they will default to\n"
        " \"-\" for standard input and standard output, respectively.\n");
}

typedef enum { IO_READ, IO_WRITE, } IO_mode;
typedef enum { OP_ENC, OP_DEC, } Operation;

typedef struct {
    int fd;                     /* file descriptor */
    IO_mode mode;
    size_t fill;                /* fill index */
    size_t read;                /* read index */
    size_t size;
    size_t total;
    uint8_t buf[];
} io_handle;

typedef struct {
    uint8_t window_sz2;
    uint8_t lookahead_sz2;
    size_t decoder_input_buffer_size;
    size_t buffer_size;
    uint8_t verbose;
    Operation cmd;
    char *in_fname;
    char *out_fname;
    io_handle *in;
    io_handle *out;
} config;

static void die(char *msg) {
    fprintf(stderr, "%s\n", msg);
    exit(EXIT_FAILURE);
}

static void report(config *cfg);

/* Open an IO handle. Returns NULL on error. */
static io_handle *handle_open(char *fname, IO_mode m, size_t buf_sz) {
    io_handle *io = NULL;
    io = malloc(sizeof(*io) + buf_sz);
    if (io == NULL) { return NULL; }
    memset(io, 0, sizeof(*io) + buf_sz);
    io->fd = -1;
    io->size = buf_sz;
    io->mode = m;

    if (m == IO_READ) {
        if (0 == strcmp("-", fname)) {
            io->fd = STDIN_FILENO;
        } else {
            io->fd = open(fname, O_RDONLY | O_BINARY);
        }
    } else if (m == IO_WRITE) {
        if (0 == strcmp("-", fname)) {
            io->fd = STDOUT_FILENO;
        } else {
            io->fd = open(fname, O_WRONLY | O_BINARY | O_CREAT | O_TRUNC /*| O_EXCL*/, 0644);
        }
    }

    if (io->fd == -1) {         /* failed to open */
        free(io);
        HEATSHRINK_ERR(1, "open");
        return NULL;
    }

    return io;
}

/* Read SIZE bytes from an IO handle and return a pointer to the content.
 * BUF contains at least size_t bytes. Writes the number of bytes read
 * into *read_length. Returns true on success, false on error. */
static bool
handle_read(io_handle *io, size_t size, uint8_t **buf, size_t *read_size) {
    LOG("@ read %zd\n", size);
    if (buf == NULL || read_size == NULL) { return false; }
    if (size > io->size) {
        fprintf(stderr, "size %zd, io->size %zd\n", size, io->size);
        return false;
    }
    if (io->mode != IO_READ) { return false; }

    size_t rem = io->fill - io->read;
    if (rem >= size) {
        *buf = &io->buf[io->read];
        *read_size = size;
        return true;
    } else {                    /* read and replenish */
        if (io->fd == -1) {     /* already closed, return what we've got */
            *buf = &io->buf[io->read];
            *read_size = rem;
            return true;
        }

        memmove(io->buf, &io->buf[io->read], rem);
        io->fill -= io->read;
        io->read = 0;
        ssize_t read_sz = read(io->fd, &io->buf[io->fill], io->size - io->fill);
        if (read_sz < 0) {
            HEATSHRINK_ERR(1, "read");
            return false;
        }
        io->total += (size_t)read_sz;
        if (read_sz == 0) {     /* EOF */
            if (close(io->fd) < 0) {
                HEATSHRINK_ERR(1, "close");
                return false;
            }
            io->fd = -1;
        }
        io->fill += (size_t)read_sz;
        *buf = io->buf;
        *read_size = (io->fill > size ? size : io->fill);
        return true;
    }
}

/* Drop the oldest SIZE bytes from the buffer. Returns false on error. */
static bool
handle_drop(io_handle *io, size_t size) {
    LOG("@ drop %zd\n", size);
    if (io->read + size <= io->fill) {
        io->read += size;
    } else {
        return false;
    }
    if (io->read == io->fill) {
        io->read = 0;
        io->fill = 0;
    }
    return true;
}

/* Sink SIZE bytes from INPUT into the io handle. Returns the number of
 * bytes written, or -1 on error. */
static bool
handle_sink(io_handle *io, size_t size, uint8_t *input, size_t *sink_size) {
    LOG("@ sink %zd\n", size);
    if (size > io->size) { return false; }
    if (io->mode != IO_WRITE) { return false; }
    if (sink_size == NULL) { return false; }

    if (io->fill + size > io->size) {
        ssize_t written = write(io->fd, io->buf, io->fill);
        LOG("@ flushing %zd, wrote %zd\n", io->fill, written);
        if (written < 0) { HEATSHRINK_ERR(1, "write"); }
        io->total += (size_t)written;
        memmove(io->buf, &io->buf[written], io->fill - (size_t)written);
        io->fill -= (size_t)written;
    }
    memcpy(&io->buf[io->fill], input, size);
    io->fill += size;
    *sink_size = size;
    return true;
}

static bool
handle_close(io_handle *io) {
    if (io->fd != -1) {
        if (io->mode == IO_WRITE) {
            ssize_t written = write(io->fd, io->buf, io->fill);
            if (written < 0) {
                HEATSHRINK_ERR(1, "write");
                return false;
            }
            io->total += (size_t)written;
            LOG("@ close: flushing %zd, wrote %zd\n", io->fill, written);
        }
        close(io->fd);
        io->fd = -1;
    }
    return true;
}

static void close_and_report(config *cfg) {
    (void)handle_close(cfg->in);
    (void)handle_close(cfg->out);
    if (cfg->verbose) { report(cfg); }
    free(cfg->in);
    free(cfg->out);
}

static int encoder_sink_read(config *cfg, heatshrink_encoder *hse,
        uint8_t *data, size_t data_sz) {
    size_t out_sz = cfg->buffer_size;
    uint8_t out_buf[out_sz];
    memset(out_buf, 0, out_sz);
    size_t sink_sz = 0;
    size_t poll_sz = 0;
    HSE_sink_res sres;
    HSE_poll_res pres;
    HSE_finish_res fres;
    io_handle *out = cfg->out;

    size_t sunk = 0;
    do {
        if (data_sz > 0) {
            sres = heatshrink_encoder_sink(hse, &data[sunk], data_sz - sunk, &sink_sz);
            if (sres < 0) { die("sink"); }
            sunk += sink_sz;
        }
        
        do {
            pres = heatshrink_encoder_poll(hse, out_buf, out_sz, &poll_sz);
            if (pres < 0) { die("poll"); }
            size_t sunk_bytes;
            if (!handle_sink(out, poll_sz, out_buf, &sunk_bytes)) {
                die("handle_sink");
            }
        } while (pres == HSER_POLL_MORE);
        
        if (poll_sz == 0 && data_sz == 0) {
            fres = heatshrink_encoder_finish(hse);
            if (fres < 0) { die("finish"); }
            if (fres == HSER_FINISH_DONE) { return 1; }
        }
    } while (sunk < data_sz);
    return 0;
}

static int encode(config *cfg) {
    uint8_t window_sz2 = cfg->window_sz2;
    size_t window_sz = 1 << window_sz2; 
    heatshrink_encoder *hse = heatshrink_encoder_alloc(window_sz2, cfg->lookahead_sz2);
    if (hse == NULL) { die("failed to init encoder: bad settings"); }
    size_t read_sz = 0;
    io_handle *in = cfg->in;

    /* Process input until end of stream */
    while (1) {
        uint8_t *input = NULL;
        if (!handle_read(in, window_sz, &input, &read_sz) || input == NULL) {
            fprintf(stderr, "handle read failure\n");
            die("read");
        }

        /* Pass read to encoder and check if input is fully processed. */
        if (encoder_sink_read(cfg, hse, input, read_sz)) break;

        if (!handle_drop(in, read_sz)) { die("drop"); }
    }

    heatshrink_encoder_free(hse);
    close_and_report(cfg);
    return 0;
}

static int decoder_sink_read(config *cfg, heatshrink_decoder *hsd,
        uint8_t *data, size_t data_sz) {
    io_handle *out = cfg->out;
    size_t sink_sz = 0;
    size_t poll_sz = 0;
    size_t out_sz = cfg->buffer_size;
    uint8_t out_buf[out_sz];
    memset(out_buf, 0, out_sz);

    HSD_sink_res sres;
    HSD_poll_res pres;
    HSD_finish_res fres;

    size_t sunk = 0;
    do {
        if (data_sz > 0) {
            sres = heatshrink_decoder_sink(hsd, &data[sunk], data_sz - sunk, &sink_sz);
            if (sres < 0) { die("sink"); }
            sunk += sink_sz;
        }

        do {
            pres = heatshrink_decoder_poll(hsd, out_buf, out_sz, &poll_sz);
            if (pres < 0) { die("poll"); }
            size_t sink_bytes;
            if (!handle_sink(out, poll_sz, out_buf, &sink_bytes)) {
                die("handle_sink");
            }
        } while (pres == HSDR_POLL_MORE);
        
        if (data_sz == 0 && poll_sz == 0) {
            fres = heatshrink_decoder_finish(hsd);
            if (fres < 0) { die("finish"); }
            if (fres == HSDR_FINISH_DONE) { return 1; }
        }
    } while (sunk < data_sz);

    return 0;
}

static int decode(config *cfg) {
    uint8_t window_sz2 = cfg->window_sz2;
    size_t window_sz = 1 << window_sz2;
    size_t ibs = cfg->decoder_input_buffer_size;
    heatshrink_decoder *hsd = heatshrink_decoder_alloc((uint16_t)ibs,
        window_sz2, cfg->lookahead_sz2);
    if (hsd == NULL) { die("failed to init decoder"); }

    io_handle *in = cfg->in;

    HSD_finish_res fres;

    /* Process input until end of stream */
    while (1) {
        uint8_t *input = NULL;
        size_t read_sz = 0;
        if (!handle_read(in, window_sz, &input, &read_sz) || input == NULL) {
            fprintf(stderr, "handle read failure\n");
            die("read");
        }
        if (read_sz == 0) {
            fres = heatshrink_decoder_finish(hsd);
            if (fres < 0) { die("finish"); }
            if (fres == HSDR_FINISH_DONE) break;
        } else {
            if (decoder_sink_read(cfg, hsd, input, read_sz)) { break; }
            if (!handle_drop(in, read_sz)) { die("drop"); }
        }
    }
        
    heatshrink_decoder_free(hsd);
    close_and_report(cfg);
    return 0;
}

static void report(config *cfg) {
    size_t inb = cfg->in->total;
    size_t outb = cfg->out->total;
    fprintf(cfg->out->fd == STDOUT_FILENO ? stderr : stdout,
        "%s %0.2f %%\t %zd -> %zd (-w %u -l %u)\n",
        cfg->in_fname, 100.0 - (100.0 * (double)outb) / (double)inb, inb, outb,
        cfg->window_sz2, cfg->lookahead_sz2);
}

static void proc_args(config *cfg, int argc, char **argv) {
    cfg->window_sz2 = DEF_WINDOW_SZ2;
    cfg->lookahead_sz2 = DEF_LOOKAHEAD_SZ2;
    cfg->buffer_size = DEF_BUFFER_SIZE;
    cfg->decoder_input_buffer_size = DEF_DECODER_INPUT_BUFFER_SIZE;
    cfg->cmd = OP_ENC;
    cfg->verbose = 0;
    cfg->in_fname = "-";
    cfg->out_fname = "-";

    int a = 0;
    while ((a = getopt(argc, argv, "hedi:w:l:v")) != -1) {
        switch (a) {
        case 'h':               /* help */
            usage();
            exit(EXIT_SUCCESS);
        case 'e':               /* encode */
            cfg->cmd = OP_ENC; break;
        case 'd':               /* decode */
            cfg->cmd = OP_DEC; break;
        case 'i':               /* input buffer size */
            cfg->decoder_input_buffer_size = (size_t)atoi(optarg);
            break;
        case 'w':               /* window bits */
            cfg->window_sz2 = (uint8_t)atoi(optarg);
            if (cfg->window_sz2 < HEATSHRINK_MIN_WINDOW_BITS
                || cfg->window_sz2 > HEATSHRINK_MAX_WINDOW_BITS) {
                fprintf(stderr, "Invalid window bits: %u\n", cfg->window_sz2);
                exit(EXIT_FAILURE);
            }
            break;
        case 'l':               /* lookahead bits */
            cfg->lookahead_sz2 = (uint8_t)atoi(optarg);
            break;
        case 'v':               /* verbosity++ */
            cfg->verbose++;
            break;
        case '?':               /* unknown argument */
        default:
            usage();
            exit(EXIT_FAILURE);
        }
    }
    argc -= optind;
    argv += optind;
    if (argc > 0) {
        cfg->in_fname = argv[0];
        argc--;
        argv++;
    }
    if (argc > 0) { cfg->out_fname = argv[0]; }

    if (cfg->lookahead_sz2 < HEATSHRINK_MIN_LOOKAHEAD_BITS
        || cfg->lookahead_sz2 >= cfg->window_sz2) {
        fprintf(stderr, "Invalid lookahead bits: %u\n", cfg->lookahead_sz2);
        exit(EXIT_FAILURE);
    }
}

int main(int argc, char **argv) {
    config cfg;
    memset(&cfg, 0, sizeof(cfg));
    proc_args(&cfg, argc, argv);

    if (0 == strcmp(cfg.in_fname, cfg.out_fname)
        && (0 != strcmp("-", cfg.in_fname))) {
        fprintf(stderr, "Refusing to overwrite file '%s' with itself.\n", cfg.in_fname);
        exit(1);
    }

    cfg.in = handle_open(cfg.in_fname, IO_READ, cfg.buffer_size);
    if (cfg.in == NULL) { die("Failed to open input file for read"); }
    cfg.out = handle_open(cfg.out_fname, IO_WRITE, cfg.buffer_size);
    if (cfg.out == NULL) { die("Failed to open output file for write"); }

#ifdef _WIN32
    /*
     * On Windows, stdin and stdout default to text mode. Switch them to
     * binary mode before sending data through them.
     */
    _setmode(STDOUT_FILENO, O_BINARY);
    _setmode(STDIN_FILENO, O_BINARY);
#endif

    if (cfg.cmd == OP_ENC) {
        return encode(&cfg);
    } else if (cfg.cmd == OP_DEC) {
        return decode(&cfg);
    } else {
        usage();
        return EXIT_FAILURE;
    }
}
